File: src/Stats.cpp
#include <iostream>
#include <math.h>
#include <assert.h>
#include "Stats.h"
#include "PRNG.h"

namespace csci480 {
  Stats::Stats() { reset(); }
  void Stats::reset() { samples = 0; sum=0; sum2=0; min = NAN; max = NAN; }

  void Stats::sample(double value) {
    if (samples == 0) {
      samples = 1;
      sum = value;
      sum2 = value*value;
      min = value;
      max = value;
    } else {
      ++samples;
      sum += value;
      sum2 += value*value;
      if (value < min) min = value;
      if (value > max) max = value;
    }
  }

  double Stats::mean() const {
    return sum/samples;
  }

  double Stats::stddev() const {
    return sqrt((sum2-pow(sum,2)/samples)/(samples-1.0));
  }

  void Stats::ok() {
    Stats s;
    PRNG prng(1);
    int n = 10000;

    for (int i=0; i<n; ++i) {
      double x=prng.next(0,1000000)/1000000.0;
      s.sample(x);
    }

    assert(fabs(s.mean()-0.5)<1/sqrt(n));
    assert(fabs(s.stddev()-sqrt(1/12.0))<1/sqrt(n));    
  }
}

File: algorithms/src/main/java/csci480/Stats.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package csci480;

import static org.apache.commons.math3.special.Erf.erfcInv;

/**
 *
 * @author wmacevoy
 */
public class Stats {

    public int samples;
    public double sum, sum2, min, max;

    public Stats() {
        reset();
    }

    public void reset() {
        samples = 0;
        sum = 0;
        sum2 = 0;
        min = Double.NaN;
        max = Double.NaN;
    }

    public void sample(double value) {
        if (samples == 0) {
            samples = 1;
            sum = value;
            sum2 = value * value;
            min = value;
            max = value;
        } else {
            ++samples;
            sum += value;
            sum2 += value * value;
            if (value < min) {
                min = value;
            }
            if (value > max) {
                max = value;
            }
        }
    }

    public double mean() {
        return sum / samples;
    }

    public double stddev() {
        return Math.sqrt((sum2 - Math.pow(sum, 2) / samples) / (samples - 1.0));
    }

    static final double MIN_SAMPLES = 30.0;

    public static double requiredSamples(double relativeError, double xbar, double sigma) {
        double zMax = erfcInv(relativeError);
        double nMax = Math.pow(zMax * sigma / (xbar * relativeError), 2.0);
        return Math.max(nMax, MIN_SAMPLES);
    }

    public double requiredSamples(double relativeError) {
        if (samples < MIN_SAMPLES) {
            return MIN_SAMPLES;
        }
        return requiredSamples(relativeError, mean(), stddev());
    }

    public double requiredSamples() {
        return requiredSamples(0.05);
    }

}
